Sources/SwiftProtobuf/JSONTypeAdditions.swift

```diff
@@ -420,40 +216,25 @@ public extension Message {
         encoder.append(text: json)
     }
 
-    static func setFromJSON(decoder: inout JSONDecoder, value: inout Self?) throws {
-        if decoder.scanner.skipOptionalNull() {
-            // Fields of type google.protobuf.Value actually get set from 'null'
-            if self == Google_Protobuf_Value.self {
-                value = Self()
-            } else {
-                // All other message field types treat 'null' as an unset field
-                value = nil
-            }
-            return
-        }
-        let message = try Self(decoder: &decoder)
-        value = message
-    }
-
-    static func setFromJSON(decoder: inout JSONDecoder, value: inout [Self]) throws {
-        let message = try Self(decoder: &decoder)
-        value.append(message)
+    public init(json: String) throws {
+        let data = json.data(using: String.Encoding.utf8)!
+        try self.init(jsonUTF8: data)
     }
 
-    public init(json: String) throws {
-        var decoder = JSONDecoder(json: json)
-        if decoder.scanner.skipOptionalNull() {
-            self.init()
-        } else {
-            try self.init(decoder: &decoder)
-        }
-        if !decoder.scanner.complete {
-            throw DecodingError.trailingGarbage
+    public init(jsonUTF8: Data) throws {
+        self.init()
+        try jsonUTF8.withUnsafeBytes { (bytes:UnsafePointer<UInt8>) in
+            var decoder = JSONDecoder(utf8Pointer: bytes, count: jsonUTF8.count)
+            if !decoder.scanner.skipOptionalNull() {
+                try self.decodeJSON(from: &decoder)
+            }
+            if !decoder.scanner.complete {
+                throw JSONDecodingError.trailingGarbage
+            }
         }
     }
 
-    init(decoder: inout JSONDecoder) throws {
-        self.init()
+    public mutating func decodeJSON(from decoder: inout JSONDecoder) throws {
```

Comment:
I've managed to bring the three decoders into a similar structure:
* initializers present the user-facing API (here that's `init(json:String)` and `init(jsonUTF8:Data)`
* There's a `decode<Format>(from: inout <Format>Decoder)` on `Message` that provides the override hook for specific messages to override specific formats.  The contracts aren't entirely well-documented yet.